assets/model_monitoring/components/src/generation_safety_quality/annotation_compute_histogram/run.py

```diff
@@ -1695,73 +734,122 @@ def apply_annotation(
-            # _PromptBuilder can iteratively batch unlabeled examples
-            prompt_builder = _PromptBuilder(
-                template=ANNOTATION_TEMPLATES[metric_name],
-                template_requirements=annotation_requirements[metric_name],
-                max_tokens=max_prompt_tokens,
-                min_input_examples=1,
+        metric_name_compact = get_compact_metric_name(metric_name)
+        column_name = COMPACT_METRIC_NAME_TO_COLUMN[metric_name_compact]
+        metrics_list.append(column_name)
+    has_context = CONTEXT in production_df.columns
+    has_ground_truth = GROUND_TRUTH in production_df.columns
+
+    def annotate_batch(iterator):
+        for batch in iterator:
+            # add environment variables on executors
+            for env_var_key, env_var_value in driver_env_vars.items():
+                os.environ[env_var_key] = env_var_value
+
+            rows = []
+            for index, row in batch.iterrows():
+                qca = {PROMPT: row[PROMPT], COMPLETION: row[COMPLETION]}
+                if has_context:
+                    qca[CONTEXT] = row[CONTEXT]
+                if has_ground_truth:
+                    qca[GROUND_TRUTH] = row[GROUND_TRUTH]
+                rows.append(qca)
+
+            output_dir = tempfile.TemporaryDirectory()
+            evaluate(
+                evaluation_name="gsq-evaluation",
+                data=rows,
+                task_type="qa",
+                data_mapping={
+                    "questions": PROMPT,
+                    "contexts": CONTEXT,
+                    "y_pred": COMPLETION,
+                    "y_test": GROUND_TRUTH
+                },
+                model_config={
+                    "api_version": api_version,
+                    "api_base": api_base,
+                    "api_type": AZURE,
+                    "api_key": api_key,
+                    "deployment_id": model_type
+                },
+                metrics_list=metrics_list,
+                output_path=output_dir.name
             )
-
-            # _JobManager handles batching and retry logic
-            job_manager = _JobManager(
-                prompt_builder=prompt_builder,
-            )
-
-            for batch in iterator:
-                # add environment variables on executors
-                for env_var_key, env_var_value in driver_env_vars.items():
-                    os.environ[env_var_key] = env_var_value
-
-                print("Copied environment variables from driver to executor.")
-
-                yield job_manager.submit_inputs(
-                    token_manager=token_manager,
-                    input_data_df=batch,
-                    request_args=request_args,
-                    max_inputs=max_inputs,
-                    num_samples=num_samples,
-                    **endpoint_args,
-                )
-
+            tabular_result = pd.read_json(os.path.join(output_dir.name, "eval_results.jsonl"), lines=True)
```

Comment:
This code will run in spark, will it make performance better to read the data using pyspark instead of pandas

Reply:
This code is run within mapInPandas function:
https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.mapInPandas.html
so we are working on pandas dataframes here, we can't load it in spark.
Actually I would prefer to just get a pandas dataframe directly from evaluate - I gave this as feedback to Ankit - that way we wouldn't need to save and load from disk which is expensive in execution time, although compared to calling openai service many times the time is negligible.